add time_zones tables

bharatnc opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

- New Feature



Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add `time_zones` table.

Detailed description / Documentation draft:

This table adds a new table called the `time_zones` table:

```
:) select * from system.time_zones limit 10;

SELECT *
FROM system.time_zones
LIMIT 10

┌─time_zone───────────┐
│ America/Thunder_Bay │
│ Africa/Blantyre     │
│ Africa/Lusaka       │
│ America/Managua     │
│ Europe/Chisinau     │
│ Etc/GMT-7           │
│ Pacific/Midway      │
│ Asia/Irkutsk        │
│ America/Panama      │
│ Etc/GMT+5           │
└─────────────────────┘

10 rows in set. Elapsed: 0.012 sec.
```

The available timezones are parsed from under
`ClickHouse/contrib/cctz/testdata/zoneinfo` through a script
and the table content is constructed from the
`StorageSystemTimeZones.generated.cpp`.

I was able to test this locally and was able to get a list of time
zones as shown above.

This Addresses: #12444

bharatnc committed:
add time_zones tables

This table adds a new table called the `time_zones` table:

```
:) select * from system.time_zones limit 10;

SELECT *
FROM system.time_zones
LIMIT 10

┌─time_zone──────────┐
│ Africa/Abidjan     │
│ Africa/Accra       │
│ Africa/Addis_Ababa │
│ Africa/Algiers     │
│ Africa/Asmara      │
│ Africa/Asmera      │
│ Africa/Bamako      │
│ Africa/Bangui      │
│ Africa/Banjul      │
│ Africa/Bissau      │
└────────────────────┘

10 rows in set. Elapsed: 0.012 sec.
```

The available timezones are parsed from under
`ClickHouse/contrib/cctz/testdata/zoneinfo` through a script
and the table content is constructed from the
StorageSystemTimeZones.generated.cpp`.

I was able to test this locally and was able to get a list of time
zones as shown above.

alexey-milovidov commented on src/Storages/System/StorageSystemTimeZones.cpp:
```diff
@@ -0,0 +1,27 @@
+
+void StorageSystemTimeZones::fillData(MutableColumns & res_columns, const Context &, const SelectQueryInfo &) const
+{
+    std::vector<const char *> time_zones;
```
Maybe we can do it in single loop, without temporary array?

bharatnc replied:
yes definitely, should be fixed now.

alexey-milovidov reviewed (commented):
There is one slight deficiency in this implementation.

So, the list of timezones is generated at release time and committed to sources.
But when we build ClickHouse without a release, the list of timezones actually used may be different.
Instead, we can obtain the list of timezones during the build and generate source file in build directory.

Why it was done that way for system.contributors? Because we cannot obtain the list of contributors directly from source tree during build process. It requires git. But git is not necessary available during build - for example, if we build from source tarballs. So, for system.contributors it is better to pre-generate data.

But for system.time_zones, we don't have to pre-generate data. It's better to get the actual list of timezones that was used during build. See `contrib/cctz-cmake/CMakeLists.txt`

bharatnc committed:
attempt to generate tz data during build using cmake

bharatnc committed:
add tests for new time_zones table

bharatnc committed:
remove older script that generated tz data

bharatnc committed:
fix vertical spacing

bharatnc commented:
> There is one slight deficiency in this implementation.
> 
> So, the list of timezones is generated at release time and committed to sources.
> But when we build ClickHouse without a release, the list of timezones actually used may be different.
> Instead, we can obtain the list of timezones during the build and generate source file in build directory.
> 
> Why it was done that way for system.contributors? Because we cannot obtain the list of contributors directly from source tree during build process. It requires git. But git is not necessary available during build - for example, if we build from source tarballs. So, for system.contributors it is better to pre-generate data.
> 
> But for system.time_zones, we don't have to pre-generate data. It's better to get the actual list of timezones that was used during build. See `contrib/cctz-cmake/CMakeLists.txt`

1. I have refactored the PR to directly generate the `StorageSystemTimeZones.generated.cpp` file from within the `cctz-cmake/CMakeList.txt` file.

2. For a OS_LINUX we are now using all the time-zones that's the same as the ones that's being baked into the ClickHouse binary.

3. Regarding 2 - this also means that this list would not work for other systems like OS_X, BSD etc. I'm checking for the presence of the generated`StorageSystemTimeZones.generated.cpp` file before attaching the `system.time_zones` table which means, this table won't exist currently for other OSes except OS_LINUX. Do you recommend a work around here or this sounds good to you?  Thank you for your feedback.

bharatnc committed:
fix more spacing

alexey-milovidov commented:
> Do you recommend a work around here or this sounds good to you?

I think it's Ok.

bharatnc committed:
include reference file for test

bharatnc committed:
spacing, new line at end in .sql and .reference files

bharatnc commented:
> Looks like you forgot to commit test reference file.

oops, yes, just added.

filimonov commented:
Just a reminder / related - i think that can be merged https://github.com/ClickHouse/ClickHouse/pull/13648

bharatnc commented:
> > Looks like you forgot to commit test reference file.
> 
> oops, yes, just added.

Test fails with:
```bash
2020-08-20 22:32:54 
2020-08-20 22:32:54 Received exception from server (version 20.8.1):
2020-08-20 22:32:54 Code: 60. DB::Exception: Received from localhost:9000. DB::Exception: Table system.time_zones doesn't exist.. 
2020-08-20 22:32:54 
2020-08-20 22:32:54 , result:
2020-08-20 22:32:54 
2020-08-20 22:32:54 
```

Will figure out what's going on - not sure if the table get's attached in the way I expect it to.

bharatnc commented:
The build uses the binary: [clickhouse-server_20.8.1.4436_all.deb](https://clickhouse-builds.s3.yandex.net/13880/1cb9fe21b911c6bccfaddad1a00306927da9a8ac/clickhouse_build_check/gcc-9_relwithdebuginfo_none_bundled_unsplitted_disable_False_deb/clickhouse-server_20.8.1.4436_all.deb) which I downloaded and installed locally on my linux dev machine and I'm able to run that failing query just fine:

```bash
:) SELECT if ((SELECT count(*) FROM system.time_zones) = 562, 'ok', 'fail');

SELECT if(
    (
        SELECT count(*)
        FROM system.time_zones
    ) = 562, 'ok', 'fail')

┌─if(equals(_subquery1, 562), 'ok', 'fail')─┐
│ ok                                        │
└───────────────────────────────────────────┘
```

Not entirely sure why it's failing.

bharatnc committed:
commit StorageSystemTimeZones.generated.cpp

bharatnc committed:
try and fix gcc-9 build.

This tries to fix the failing gcc-9 unbundled build:

https://clickhouse-builds.s3.yandex.net/13880/e437d85b927a47da69b8bdc4dde95917c2261cbe/clickhouse_build_check/build_log_753740668_1598037713.txt

bharatnc committed:
Docs for system.time_zones table

alexey-milovidov commented:
Stress test (undefined) - fixed in #13980
Performance - sporadic changes, Ok.

alexey-milovidov commented:
There a few minor changes required... I will add them in another PR and this PR will be merged automatically.

alexey-milovidov commented on tests/queries/0_stateless/01455_time_zones.sql:
```diff
@@ -0,0 +1,9 @@
+-- ┌─count()─┐
+-- │     562 │
+-- └─────────┘
+SELECT if ((SELECT count(*) FROM system.time_zones) = 562, 'ok', 'fail');
```
In recent commit we started to generate the complete list on the fly. And the number of timezones may change. I will just check that there are more than 500 :)

alexey-milovidov commented on src/Storages/ya.make:
```diff
@@ -182,6 +182,8 @@ SRCS(
     System/StorageSystemTableFunctions.cpp
     System/StorageSystemTables.cpp
+    System/StorageSystemTimeZones.cpp
+    System/StorageSystemTimeZones.generated.cpp
```
Actually ya.make is used for "alien build system" that are using timezones in completely different way... I will exclude system.timezones for that build.

alexey-milovidov commented on src/Storages/System/StorageSystemTimeZones.generated.cpp:
```diff
@@ -0,0 +1,565 @@
+// autogenerated by ClickHouse/contrib/cctz-cmake/CMakeLists.txt
```
This should not be committed.

bharatnc replied:
I see -  interesting, I realized that the tests fail with `system.time_zones` table not found without committing this file. Besides files like the [StorageSystemContributors.generated.cpp](https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/System/StorageSystemContributors.generated.cpp), [StorageSystemBuildOptions.generated.cpp.in](https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/System/StorageSystemBuildOptions.generated.cpp.in) are being committed though. 

Though I realized that downloading the exact binary being build and running locally actually shows the systems.time_zones table all fine.

alexey-milovidov commented on contrib/cctz-cmake/CMakeLists.txt:
```diff
@@ -24,8 +24,14 @@ if (USE_INTERNAL_CCTZ)
+    # StorageSystemTimeZones.generated.cpp is autogenerated each time during a build
+    # data in this file will be used to populate the system.time_zones table, this is specific to OS_LINUX
+    # as the library that's built using embedded tzdata is also specific to OS_LINUX
+    set(SYSTEM_STORAGE_TZ_FILE "../../src/Storages/System/StorageSystemTimeZones.generated.cpp")
```
I will try to use ${CMAKE_CURRENT_BINARY_DIR} just in case...

alexey-milovidov commented:
Merged in #14030.
